### PR TITLE
#162949040 Users should be able to create and view articles

### PR DESCRIPTION
**Pull request by jomasim**
#### What does this PR do?

Enables a user to create, view single and all articles.

#### Description of Task to be completed?
- Create article listing page
- Consume articles API endpoint
- Write tests for the feature

#### How should this be manually tested?
1. Navigate to a folder where you would prefer to test the repository.

2. Create a directory
 ```$```  ```mkdir myproject```

3. Navigate to the directory "myproject"
 ```$```   ```cd myproject```
4. Clone the author's haven repository
  ```$```   ```git clone https://github.com/andela/ah-robotics-frontend.git```
5. Change the directory into the "ah-robotics-fronted"
  ```$```   ```cd ah-robotics-frontend```

6.  Checkout to the branch ft-articles-162949040
 ```$```   ```git checkout ft-articles-162949040```

7. Install dependancies from the package.json
 ```$```   ```npm install``` to install all the modules listed in the `package.json` file.

8. Start the application in development mode
  ```$```   ```npm start``` to run the app in the development mode 

The application opens http://localhost:3000 in the browser.
Change the URL on the browser to;
 ```
https://ah-robotics-frontend-sta-pr-16.herokuapp.com
```
9. Click the ```Login``` link on the Navbar and enter the following details to log in:
 **(for demo purpose use)**

``` 
Username:mandela@gmail.com 
Password:happy@1234
```
### Home page lists all articles 
screenshot 2019-02-22 at 13 24 24

10. Click on the article title to view a single article
Article view page
screenshot 2019-02-22 at 13 22 45

11. Click on the `New Article` lin on the navbar to create a new article
Post Article page
screenshot 2019-02-22 at 13 23 23

###  Running Tests
```
npm test
```

#### What are the relevant pivotal tracker stories?
[#162949040](https://www.pivotaltracker.com/n/projects/2235171/stories/162949040)

### CODE REVIEW

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @jomasim,
This is good. It is working as expected. I was able to create an article and see it when I view all articles.

**Comment by waracci** [comment]
You should work on redirecting a user to the article page after the article is created.

**Comment by sharkdevs** [comment]
It is working, though there is a lot of loading and refresh. Maybe because of Heroku. Otherwise, satisfactory

File: src/components/Article/index.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import {
+ Container, Header, Image, Segment,
+} from 'semantic-ui-react';
+import renderHtml from 'react-render-html';
+import PropTypes from 'prop-types';
+
+
+const ArticleComponent = ({ article, isFetching }) => {
+  const fetchedArticle = article.article;
```
Comment (Collins33) [inline src/components/Article/index.js:12]: Use destructuring here

File: src/components/Article/index.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  return (
+    <Segment loader={isFetching} basic>
+      <Container text style={{ marginTop: '7em' }}>
+        <Header as="h1">{fetchedArticle && fetchedArticle.title}</Header>
+        <Image
+          src={(fetchedArticle && fetchedArticle.author.image)
```
Comment (Collins33) [inline src/components/Article/index.js]: can we have a variable to store values like `fetchedArticle.author.image`?

File: src/components/Article/index.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        || 'https://source.unsplash.com/random'}
+          style={{ marginTop: '2em' }}
```
Comment (Collins33) [inline src/components/Article/index.js]: Why is the styling here and not in a css file?

File: src/components/ArticleEditor/article.editor.scss [hunk 4]
```diff
@@ -0,0 +1,11 @@
+.right-column{
+ text-align: right;
+}
+.cke_contents .cke_bottom {
```
Comment (Collins33) [inline src/components/ArticleEditor/article.editor.scss]: what is `cke`?

File: src/components/ArticleEditor/index.jsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+const ArticleEditor = ({
+ onInputChange, onEditorChange, handleArticlePost, createArticle, body,
+}) => (
+  <Container>
+    <Fragment>
+      <Grid>
+        <Grid.Row>
+          <Grid.Column width={12}>
+            <p
+              style={{ color: 'red' }}
```
Comment (Collins33) [inline src/components/ArticleEditor/index.jsx]: Why are this styles here?

File: src/components/ArticlesListing/ArticleItem.jsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import { Item, Label } from 'semantic-ui-react';
+import renderHTML from 'react-render-html';
+import PropTypes from 'prop-types';
+import { NavLink } from 'react-router-dom';
+
+const ArticleItem = ({ article }) => ({ article }
+  && (
+  <Item>
+    <Item.Image src={(article.image) ? article.image : 'https://source.unsplash.com/random'} />
```
Comment (Collins33) [inline src/components/ArticlesListing/ArticleItem.jsx]: Is this link for the image a placeholder?

File: src/components/ArticlesListing/index.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import {
+  Grid, Segment, Item,
+} from 'semantic-ui-react';
+import './listings.scss';
+import PropTypes from 'prop-types';
+import ArticleItem from './ArticleItem';
+
+const ArticlesListing = ({ articles, isFetching }) => {
+  const allArticles = articles.articles;
```
Comment (Collins33) [inline src/components/ArticlesListing/index.js:10]: Use destructuring here

File: src/redux/actions/ArticleActions/articles.action.types.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+export const ARTICLE_FETCH = 'ARTICLE_FETCH';
```
Comment (Collins33) [inline src/redux/actions/ArticleActions/articles.action.types.js]: Why did you name the file as articles.action.types.js?

File: src/redux/actions/ArticleActions/articles.actions.test.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import {
+ articleFetch, articleFetchError, articleFetchSuccess, postArticle,
+  postArticleError, postArticleSuccess,
+} from './articles.actions';
+import * as types from './articles.action.types';
+
+describe('articles action tests', () => {
+it('test article fetch action', () => {
+  expect(articleFetch({}).type).toBe(types.ARTICLE_FETCH);
```
Comment (Collins33) [inline src/redux/actions/ArticleActions/articles.actions.test.js]: I expect you to test more than just the action type

File: src/redux/reducers/ArticlesReducer/fetch.articles.reducer.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+const fetchArticlesReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case types.ARTICLE_FETCH: {
+      return {
+        ...state, ...action.payload, isFetching: true,
+      };
+  }
+    case types.ARTICLE_FETCH_SUCCESS: {
+      return {
+ ...state, ...action.payload, isFetching: false,
```
Comment (Collins33) [inline src/redux/reducers/ArticlesReducer/fetch.articles.reducer.js]: Should this appear like this?

File: src/redux/sagas/ArticleSaga/postArticleSaga.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  try {
+    const response = yield call(apiClient.createUserArticle, { article: payload });
+    yield put({
+      type: types.ARTICLE_POST_SUCCESS,
+      payload: { data: response.data },
+    });
+  } catch (error) {
+    yield put({
+      type: types.ARTICLE_POST_ERROR,
+      payload: { errors: error.response ? error.response.data.article.errors : { message: 'something went wrong' } },
```
Comment (Collins33) [inline src/redux/sagas/ArticleSaga/postArticleSaga.js]: What exact error message should we use here?

**Review by Collins33 — CHANGES_REQUESTED** [review]

File: src/redux/reducers/ArticlesReducer/fetch.articles.reducer.js (hunk 10 above)
Comment (jomasim) [reply src/redux/reducers/ArticlesReducer/fetch.articles.reducer.js]: I have refactored the code

**Comment by jomasim** [comment]
@Collins33 I have managed to resolve all the issues.

**Comment by Collins33** [comment]
When I test the application, I have to reload the page and this logs me out.

**Comment by Collins33** [comment]
When I add an article, it should redirect me to another page on the application

**Comment by jomasim** [comment]
> When I add an article, it should redirect me to another page on the application

I can now redirect to the article page.

**Comment by jomasim** [comment]
> When I test the application, I have to reload the page and this logs me out.

I have fixed the issue. The app now logs out only after someone clicks log out on the navbar.

**Comment by Teatoller** [comment]
Hello @jomasim ,
I have tested the article feature and have been able to post an article.
The challenge @sharkdevs mentioned about slow loading images is an area of improvement.
I also ran the test and would advise

Regards
Steven Ennis

screenshot 2019-02-24 at 09 22 44
 you correct for a failing snapshot test.

File: src/views/CreateArticleView/index.jsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    description: '',
+    body: '',
+    errors: {},
+    tagList: [],
+  };
+
+  componentDidUpdate(prevProps, prevState, snapshot) {
+    const { history, createArticle } = this.props;
+    const { data, isPosted } = createArticle;
+    if (isPosted) {
```
Comment (Collins33) [inline src/views/CreateArticleView/index.jsx:21]: Why is there a timeout here?

**Review by Collins33 — CHANGES_REQUESTED** [review]
When I post a long article, all of it appears on the home page. This makes the home page look weird.

**Review by Teatoller — COMMENTED** [review]
Hello @jomasim 
I did make a review asking that you refined a snapshot test that is failing.
Kindly look at the comment section.

Regards
Steven Ennis

File: src/views/CreateArticleView/index.jsx (hunk 12 above)
Comment (jomasim) [reply src/views/CreateArticleView/index.jsx:21]: I have managed to add a timeout as well as added a function to truncate the text on the homepage for long paragraphs.

**Comment by jomasim** [comment]
> Hello @jomasim
> I did make a review asking that you refined a snapshot test that is failing.
> Kindly look at the comment section.
> 
> Regards
> Steven Ennis

@Teatoller I have fixed the failing test

**Review by johnwayodi — COMMENTED** [review]
@jomasim this is great work. I have tested and all the functionality works as expected.

**Review by Collins33 — APPROVED** [review]

**Pull request merged**